Proofreading MO Guide (#11605)
openvinotoolkit/openvino: docs/MO_DG/prepare_model/convert_model/mxnet_specific/Convert_Style_Transfer_From_MXNet.md

# Converting an MXNet Style Transfer Model {#openvino_docs_MO_DG_prepare_model_convert_model_mxnet_specific_Convert_Style_Transfer_From_MXNet}

This article provides instructions on how to generate a model for style transfer, using the public MXNet neural style transfer sample.

**Step 1**: Download or clone the repository [Zhaw's Neural Style Transfer repository](https://github.com/zhaw/neural_style) with an MXNet neural style transfer sample.

**Step 2**: Prepare the environment required to work with the cloned repository:

> **NOTE**: Python-tk installation is needed only for Linux. Python for Windows includes it by default.

1. Install packages dependency.<br>
```sh
sudo apt-get install python-tk
```

2. Install Python requirements:
```sh
pip3 install --user mxnet
pip3 install --user matplotlib
pip3 install --user scikit-image
```

**Step 3**: Download the pretrained [VGG19 model](https://github.com/dmlc/web-data/raw/master/mxnet/neural-style/model/vgg19.params) and save it to the root directory of the cloned repository. The sample expects the model `vgg19.params` file to be in that directory.<br>

> **Step 4**: Modify source code files of style transfer sample from cloned repository:<br>

Reviewer: ```suggestion
**Step 4**: Modify source code files of style transfer sample from the cloned repository:<br>
```

1. Go to the `fast_mrf_cnn` subdirectory.
```sh
cd ./fast_mrf_cnn
```

> 2. Open the `symbol.py` file and modify the `decoder_symbol()` function. You should see below code there:

Reviewer: ```suggestion
2. Open the `symbol.py` file and modify the `decoder_symbol()` function. You should see the following code there:
```

# Convert MXNet Style Transfer Model {#openvino_docs_MO_DG_prepare_model_convert_model_mxnet_specific_Convert_Style_Transfer_From_MXNet}

> This article provides instructions on how to generate a model for style transfer using the public MXNet neural style transfer sample.

Author: ```suggestion
This article provides instructions on how to generate a model for style transfer, using the public MXNet neural style transfer sample.
```

**Step 1**: Download or clone the repository [Zhaw's Neural Style Transfer repository](https://github.com/zhaw/neural_style) with an MXNet neural style transfer sample.

**Step 2**: Prepare the environment required to work with the cloned repository:

> **NOTE**: Python-tk installation is needed only for Linux. Python for Windows includes it by default.

1. Install packages dependency.<br>
```sh
sudo apt-get install python-tk
```

2. Install Python requirements:
```sh
pip3 install --user mxnet
pip3 install --user matplotlib
pip3 install --user scikit-image
```

> **Step 3**: Download the pre-trained [VGG19 model](https://github.com/dmlc/web-data/raw/master/mxnet/neural-style/model/vgg19.params) and save it to the root directory of the cloned repository. The sample expects the model *`vgg19.params`* file to be in that directory.<br>

Author: ```suggestion
**Step 3**: Download the pretrained [VGG19 model](https://github.com/dmlc/web-data/raw/master/mxnet/neural-style/model/vgg19.params) and save it to the root directory of the cloned repository. The sample expects the model *`vgg19.params`* file to be in that directory.<br>
```
I guess we all need to decide whether "pre-train" or "pretrain".  I'm an advocate of using the latter ;)

Reviewer 2: I'm for "pre-train" but let's wait for rest to voice their preferences :)

Author: Exactly ;)

**Step 4**: Modify source code files of style transfer sample from cloned repository:<br>

1. Go to the *`fast_mrf_cnn`* subdirectory.
```sh
cd ./fast_mrf_cnn
```

2. Open the *`symbol.py`* file and modify the *`decoder_symbol()`* function. You should see below code there:
```py
def decoder_symbol():
    data = mx.sym.Variable('data')
    data = mx.sym.Convolution(data=data, num_filter=256, kernel=(3,3), pad=(1,1), stride=(1, 1), name='deco_conv1')
```

> Replace above code with following:<br>

Author: ```suggestion
Replace the code above with the following:<br>
```

```py
def decoder_symbol_with_vgg(vgg_symbol):
    data = mx.sym.Convolution(data=vgg_symbol, num_filter=256, kernel=(3,3), pad=(1,1), stride=(1, 1), name='deco_conv1')
```

3. Save and close the *`symbol.py`* file.

4. Open and edit the *`make_image.py`* file. Go to the *`__init__()`* function in the *`Maker`* class:<br>
```py
decoder = symbol.decoder_symbol()
```
Modfiy it with the following code:<br>
```py
decoder = symbol.decoder_symbol_with_vgg(vgg_symbol)
```

> 5. To join the pre-trained weights with the decoder weights, make the following changes:

Author: ```suggestion
5. To join the pretrained weights with the decoder weights, make the following changes:
```

   After the code lines for loading the decoder weights:<br>
   ```py
   args = mx.nd.load('%s_decoder_args.nd'%model_prefix)
   auxs = mx.nd.load('%s_decoder_auxs.nd'%model_prefix)
   ```

   Add the following line:<br>
   ```py
   arg_dict.update(args)
   ```

> 6. Use *`arg_dict`* instead of *`args`* as a parameter of the *`decoder.bind()`* function. Find line below:<br>

Author: ```suggestion
6. Use *`arg_dict`* instead of *`args`* as a parameter of the *`decoder.bind()`* function. Find the line below:<br>
```

```py
self.deco_executor = decoder.bind(ctx=mx.gpu(), args=args, aux_states=auxs)
```
Replace it with the following:<br>
```py
self.deco_executor = decoder.bind(ctx=mx.cpu(), args=arg_dict, aux_states=auxs)
```
7. Add the following code to the end of the *`generate()`* function in the *`Maker`* class to save the result model as a *`.json`* file:<br>
```py
self.vgg_executor._symbol.save('{}-symbol.json'.format('vgg19'))
self.deco_executor._symbol.save('{}-symbol.json'.format('nst_vgg19'))
```
8. Save and close the *`make_image.py`* file.

**Step 5**: Follow the instructions from the *`README.md`* file in the *`fast_mrf_cnn`* directory of the cloned repository and run the sample with a decoder model.

> For example, use the following code to run the sample with the pre-trained decoder weights from the *`models`* folder and output shape:<br>

Author: ```suggestion
For example, use the following code to run the sample with the pretrained decoder weights from the *`models`* folder and output shape:<br>
```

```py
import make_image
maker = make_image.Maker('models/13', (1024, 768))
maker.generate('output.jpg', '../images/tubingen.jpg')
```

> The *`models/13`* string in the above code is composed of the following substrings:

Author: ```suggestion
The *`models/13`* string in the code above is composed of the following substrings:
```

> * *`models/`* -- path to the folder that contains *`.nd`* files with pre-trained styles weights.

Author: ```suggestion
* *`models/`* -- path to the folder that contains *`.nd`* files with pretrained styles weights.
```

* *`13`* -- prefix pointing to the default decoder for the repository, *`13_decoder`*.

> > **NOTE**: If an error prompts announcing that there is "No module named *`cPickle`*", try running the script from Step 5 in Python 2. After that return to Python 3 for the remaining steps.

Reviewer 2: ```suggestion
> **NOTE**: If "No module named *`cPickle`*"error prompts, try running the script from Step 5 in Python 2. After that return to Python 3 for the remaining steps.
```

> Any style can be selected from [collection of pre-trained weights](https://pan.baidu.com/s/1skMHqYp). On the Chinese-language page, click the down arrow next to a size in megabytes. Then wait for an overlay box to appear, and click the blue button in it to download. The *`generate()`* function generates *`nst_vgg19-symbol.json`* and *`vgg19-symbol.json`* files for the specified shape. In the code, it is [1024 x 768] for a 4:3 ratio. You can specify another, for example, [224,224] for a square ratio.

Author: ```suggestion
Any style can be selected from [collection of pretrained weights](https://pan.baidu.com/s/1skMHqYp). On the Chinese-language page, click the down arrow next to a size in megabytes. Then wait for an overlay box to appear, and click the blue button in it to download. The *`generate()`* function generates *`nst_vgg19-symbol.json`* and *`vgg19-symbol.json`* files for the specified shape. In the code, it is [1024 x 768] for a 4:3 ratio. You can specify another, for example, [224,224] for a square ratio.
```

**Step 6**: Run the Model Optimizer to generate an Intermediate Representation (IR):

1. Create a new directory. For example:<br>
```sh
mkdir nst_model
```

> 2. Copy the initial and generated model files to the created directory. For example, to copy the pre-trained decoder weights from the *`models`* folder to the *`nst_model`* directory, run the following commands:<br>

Author: ```suggestion
2. Copy the initial and generated model files to the created directory. For example, to copy the pretrained decoder weights from the *`models`* folder to the *`nst_model`* directory, run the following commands:<br>
```
